components/suggest/src/util.rs

```diff
@@ -0,0 +1,1517 @@
+///         Ok(Some(vec![chunk.to_uppercase()]))
+///     }
+/// });
+/// assert_eq!(paths.unwrap(), vec![
+///     vec!["A", "B", "C"],
+///     vec!["A", "B C"],
+///     vec!["A B C"],
+/// ]);
+/// ```
+///
+/// Mapping each chunk to multiple values:
+///
+/// ```
+/// # use suggest::util::filter_map_chunks;
+/// let paths = filter_map_chunks("a b c", 3, |chunk, _, _| {
+///     Ok(Some(vec![format!("{chunk}0"), format!("{chunk}1")]))
+/// });
+/// assert_eq!(paths.unwrap(), vec![
+///     vec!["a0", "b0", "c0"],
+///     vec!["a0", "b0", "c1"],
+///     vec!["a0", "b1", "c0"],
+///     vec!["a0", "b1", "c1"],
+///     vec!["a0", "b c0"],
+///     vec!["a0", "b c1"],
+///     vec!["a1", "b0", "c0"],
+///     vec!["a1", "b0", "c1"],
+///     vec!["a1", "b1", "c0"],
+///     vec!["a1", "b1", "c1"],
+///     vec!["a1", "b c0"],
+///     vec!["a1", "b c1"],
+///     vec!["a b0", "c0"],
+///     vec!["a b0", "c1"],
+///     vec!["a b1", "c0"],
+///     vec!["a b1", "c1"],
+///     vec!["a b c0"],
+///     vec!["a b c1"]
+/// ]);
+/// ```
+pub fn filter_map_chunks<T: Clone>(
+    query: &str,
+    max_chunk_size: usize,
+    f: impl Fn(&str, usize, &[T]) -> Result<Option<Vec<T>>>,
+) -> Result<Vec<Vec<T>>> {
+    let words: Vec<_> = query.split_whitespace().filter(|s| !s.is_empty()).collect();
+    let normalized_query = words.join(" ");
+    filter_map_chunks_recurse(
+        &words,
+        &normalized_query,
+        &mut vec![],
+        0,
+        max_chunk_size,
+        &f,
+    )
+}
+
+/// `remaining_words` is the slice of remaining words in the query string at
+/// this step. `remaining_query` is the remaining slice of the query string at
+/// this step.
+///
+/// `path` is the sequence of values returned by the filter-map function so far
+/// at this step.
+///
+/// `chunk_index` is the word-based index in the query string at this step.
+fn filter_map_chunks_recurse<T: Clone>(
+    remaining_words: &[&str],
+    remaining_query: &str,
+    path: &mut Vec<T>,
+    chunk_index: usize,
+    max_chunk_size: usize,
+    f: &impl Fn(&str, usize, &[T]) -> Result<Option<Vec<T>>>,
+) -> Result<Vec<Vec<T>>> {
+    // Filtered-in (non-pruned) paths that will be returned from this step of
+    // the traversal.
+    let mut this_step_paths: Vec<Vec<T>> = vec![];
+
+    for chunk_size in 1..=max_chunk_size {
+        if remaining_words.len() < chunk_size {
+            // `chunk_size` and the later chunk sizes in this for-loop are too
+            // big to visit the remaining words. We already visited them earlier
+            // in the loop when the chunk size was small enough.
+            break;
+        }
+
+        // Get the current chunk within the remaining query. Its char length is
+        // the sum of the lengths of the words in the chunk + `chunk_size - 1`
+        // spaces between the words.
+        let chunk_char_len = remaining_words[..chunk_size]
+            .iter()
+            .fold(chunk_size - 1, |memo, w| memo + w.len());
```

Comment:
I think this would this break if there were multiple spaces between words, but it's not since you normalize the whitespace in the parent function.  Couple thoughts on this:

- Add a docstring saying why this is safe.
- Add a test for multiple spaces below (if it's not there already, sorry if I missed it)
- Maybe `remaining_words` could be replaced with `remaining_word_positions: &[(usize, usize)]` where each item is the start/end position of the word in the original query.  That would avoid issues even if you didn't normalize.  If this adds complexity, it's not with it though.

Reply:
`filter_map_chunks_2()` had a query with extra spaces, but I moved it to its own test function so it's more obvious. I also added a comment.